baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/IllegalSafeLoggingArgument.java

```diff
@@ -497,4 +499,67 @@ private Safety getLambdaRequiredReturnSafety(LambdaExpressionTree tree, VisitorS
         }
         return SafetyAnnotations.getSafety(returnType.type(), state);
     }
+
+    @Override
+    public Description matchMemberReference(MemberReferenceTree tree, VisitorState state) {
+        // This is the type the reference gets "cast" into, whether through assignment, return type, argument, etc
+        Type castType = state.getTypes().findDescriptorType(ASTHelpers.getType(tree));
+        // The reference will be used as the expected type. This means:
+        //   * the safety of the return type of the expected must be "lower" than the actual safety of the reference
+        //   * the safety of the arguments of the expected must be "higher" than the actual safety of the reference
+        // i.e. "Supplier<@Safe String> f = this::method" shouldn't be allowed, when method returns unsafe
+        // Similarly, "@Consumer<@Unsafe String> f = this::method" shouldn't be allowed, when method expects to take a
+        //   safe argument (since it might log it)
+
+        Safety castTypeReturnSafety = SafetyAnnotations.getSafety(castType.getReturnType(), state);
+
+        MethodSymbol methodSymbol = ASTHelpers.getSymbol(tree);
+        Safety referenceReturnSafety = Safety.mergeAssumingUnknownIsSame(
+                // This gets the combined safety annotations of the method and any of its supers
+                SafetyAnnotations.getSafety(methodSymbol, state),
+                SafetyAnnotations.getSafety(methodSymbol.getReturnType(), state));
+
+        // The reference will get used as the cast type
+        // This means that we expect the cast type's return type safety and will use it as such
+        // If e.g. the cast type says it will return SAFE, but the reference returns UNSAFE, that's a problem
+        // On the other hand, if the cast returns UNSAFE and the reference returns SAFE, that's fine
+        if (!castTypeReturnSafety.allowsValueWith(referenceReturnSafety)) {
+            return buildDescription(tree)
```

Comment:
If we use `state.reportMatch(`, we can report problems with the return type as well as parameters in the same evaluation. It's not really the standard way we like to use error-prone, but I do find it helpful to describe all the issues at once for the safelogging check specifically

Reply:
Great idea indeed!